tests/cases/fourslash/completionsLiteralFromInferenceWithinInferredType.ts

```diff
@@ -0,0 +1,52 @@
+/// <reference path="fourslash.ts" />
+
+// @Filename: /a.tsx
+//// type Values<T> = T[keyof T];
+////
+//// type GetStates<T> = T extends { states: object } ? T["states"] : never;
+////
+//// type IsNever<T> = [T] extends [never] ? 1 : 0;
+////
+//// type GetIds<T, Gathered extends string = never> = IsNever<T> extends 1
+////   ? Gathered
+////   : "id" extends keyof T
+////   ? GetIds<Values<GetStates<T>>, Gathered | `#${T["id"] & string}`>
+////   : GetIds<Values<GetStates<T>>, Gathered>;
+////
+//// type StateConfig<
+////   TStates extends Record<string, StateConfig> = Record<
+////     string,
+////     StateConfig<any>
+////   >,
+////   TIds extends string = string
+//// > = {
+////   id?: string;
+////   initial?: keyof TStates & string;
+////   states?: {
+////     [K in keyof TStates]: StateConfig<GetStates<TStates[K]>, TIds>;
+////   };
+////   on?: Record<string, TIds | `.${keyof TStates & string}`>;
+//// };
+////
+//// declare function createMachine<const T extends StateConfig<GetStates<T>, GetIds<T>>>(
+////   config: T
+//// ): void;
+////
+//// createMachine({
+////   initial: "child",
+////   states: {
+////     child: {
+////       initial: "foo",
+////       states: {
+////         foo: {
+////           id: "wow_deep_id",
+////         },
+////       },
+////     },
+////   },
+////   on: {
+////     EV: "/*ts*/",
```

Comment:
In this test, which node gets marked as an “inference-blocked source”?

Reply:
IIRC it's everything from the marker up to the call-like expression. 

A way smaller repro is this:
[working TS@5.0.0-dev.20221220 playground](https://www.staging-typescript.org/play?ts=5.0.0-dev.20221220#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXwxAGcMAeAFQD4AKKALngG8AoeeAbQGl4t8BrEAE8cieOQC6DFmzZQA-A3JdxAblYyARgvgDho8mrYBfNUYCUDAG44swNc0Ilq0pDhxSjAGnUbYU9WwaDABEAPQAVOGhwd7G3uZqQA)
[broken TS@5.0.0-dev.20221221 playground](https://www.staging-typescript.org/play?ts=5.0.0-dev.20221221#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXwxAGcMAeAFQD4AKKALngG8AoeeAbQGl4t8BrEAE8cieOQC6DFmzZQA-A3JdxAblYyARgvgDho8mrYBfNUYCUDAG44swNc0Ilq0pDhxSjAGnUbYU9WwaDABEAPQAVOGhwd7G3uZqQA)

The problem in here is that those completions depend on the argument in which they are contained.

Reply:
> I think it’s currently decided per call argument, but you can always restructure multiple arguments into a single options bag and you’d hope that inference and completions behave the same.

Totally agree 👍 

> I wonder if we could refine the inference-blocking logic to say “infer from this node, but when you get to this particular sub-node, its type is the anti-any (assignable to and from nothing—we may already have a marker type like this, not sure)”?

On principle, I agree that the algorithm could be somehow smarter and that we should try to block as least as possible.  However, I also feel that the use case from the PR that regressed this is also quite important (see the test case [here](https://github.com/microsoft/TypeScript/pull/51770/files#diff-0af286c808e2e9007d71608cde10be7c90519ad5ff70a27b2c2386ebccb3f32a)). In there, we want to allow strings from the constraint. If we'd only block the "lea string node" then another place in the argument(s) could "fix" the inference for this type parameter and we likely wouldn't be able to provide autocomplete there. 

I'm open to brainstorming the improved blocking algorithm for the purpose of autocompletes - but perhaps that should be moved to a discussion thread? We might not be able to follow this rabbit right now.

Reply:
Well, the fix here of “just take whichever mode returns results” makes me worried that there are many other cases that are broken and won’t be fixed by this. I was most concerned that this could introduce new completions that are actually _not_ valid, but thinking about it more, that seems unlikely—results coming from `ContextFlags.None` should be accurate, and usually _too_ accurate. I agree this is probably an improvement.

Reply:
> Well, the fix here of “just take whichever mode returns results” makes me worried that there are many other cases that are broken and won’t be fixed by this.

I agree with that too. It is quite challenging to figure out those other use cases though, it's likely best to drive this based on the user needs. OTOH I think that completions are something that is often overlooked by people and many might not request improvements in this area.

>  I was most concerned that this could introduce new completions that are actually not valid, but thinking about it more, that seems unlikely

This "fallback" request matches the previous behavior - so at the very least, this shouldn't make the experience worse. The overall behavior (since 4.9) has still changed though (!) because this is now only the "fallback" request and not the primary one. 

> results coming from ContextFlags.None should be accurate, and usually too accurate.

What do you mean by "too accurate" here?

Reply:
> and this often has a tendency to “lock in” what you’ve already typed

Yeah. I will probably just reiterate over what we both already probably have in minds here, but it might be helpful for others.
1. it's all heuristics-based, overall the inference-related cases reuse the standard inference algorithm but it's influenced with flags to behave slightly differently to avoid at times "fixing"/"locking in" the inferences
2. it's rather hard (impossible?) to know what's the actual intended usage of the given autocomplete and it's hard to collect multiple class of candidates for this purpose within the inference algorithm
3. I think that we can distinguish at least two types of completions
- one that impacts the inference itself (one that might act as a direct source for that particular inference) - for those we don't want to lock the inference because it would be nice for people to be able to choose different values from the ones available in the constraint etc. And if the current available value is already a valid one then without blocking etc we wouldn't be able to access the type from the constraint easily
- the other type is when those completions are often **completely** derived from other types. They are not acting as any kind of inference sources.
4. note that the relation to the constraint might very complex so it's not that easy to distinguish between those two cases upfront
5. we could potentially achieve better results with a smarter blocking but I would still be worried that at times it could lead to unwanted results in cases like: `function t<T extends 'foo' | 'bar'>({ a: T, b: T }): void`. Even if we'd only block `b`, without blocking the parent object then it's likely that `T` would get fixed with `a`'s value and ultimately we wouldn't receive `'foo' | 'bar'` as valid completions.
6. I think it's definitely worth experimenting further with this algorithm and all - but for the time being this double request approach is "good enough" and something that we could try to iterate on. It's also nice that this PR brings some new tests into the codebase and that they will easily inform the future experiments with this algorithm about the use cases/requirements.